polkadot/node/core/prospective-parachains/src/fragment_tree.rs

```diff
@@ -756,53 +756,127 @@ impl FragmentTree {
 		depths.iter_ones().collect()
 	}
 
-	/// Select a candidate after the given `required_path` which passes
-	/// the predicate.
+	/// Select `count` candidates after the given `required_path` which pass
+	/// the predicate and have not already been backed on chain.
 	///
-	/// If there are multiple possibilities, this will select the first one.
-	///
-	/// This returns `None` if there is no candidate meeting those criteria.
+	/// Does an exhaustive search into the tree starting after `required_path`.
+	/// If there are multiple possibilities of size `count`, this will select the first one.
+	/// If there is no chain of size `count` that matches the criteria, this will return the largest
+	/// chain it could find with the criteria.
+	/// If there are no candidates meeting those criteria, returns an empty `Vec`.
+	/// Cycles are accepted, see module docs for the `Cycles` section.
 	///
 	/// The intention of the `required_path` is to allow queries on the basis of
 	/// one or more candidates which were previously pending availability becoming
 	/// available and opening up more room on the core.
-	pub(crate) fn select_child(
+	pub(crate) fn select_children(
 		&self,
 		required_path: &[CandidateHash],
+		count: u32,
 		pred: impl Fn(&CandidateHash) -> bool,
-	) -> Option<CandidateHash> {
+	) -> Vec<CandidateHash> {
 		let base_node = {
 			// traverse the required path.
 			let mut node = NodePointer::Root;
 			for required_step in required_path {
-				node = self.node_candidate_child(node, &required_step)?;
+				if let Some(next_node) = self.node_candidate_child(node, &required_step) {
+					node = next_node;
+				} else {
+					return vec![]
+				};
 			}
 
 			node
 		};
 
-		// TODO [now]: taking the first selection might introduce bias
+		// TODO: taking the first best selection might introduce bias
```

Comment:
no, this was here beforehand
It could be solved by randomising the order in which we visit a node's children, but that would make tests harder to write.
Since validators don't favour particular collators when requesting collations, the order of potential forks in the fragment tree should already be somewhat "random" based on network latency (unless collators find a way to censor/DOS other collators)